src/Lean/Elab/Command.lean

```diff
@@ -520,8 +520,12 @@ def elabCommandTopLevel (stx : Syntax) : CommandElabM Unit := withRef stx do pro
     -- recovery more coarse. In particular, If `c` in `set_option ... in $c` fails, the remaining
     -- `end` command of the `in` macro would be skipped and the option would be leaked to the outside!
     elabCommand stx
-    withLogging do
-      runLinters stx
+    -- Run the linters, unless `#guard_msgs` is present, which is special and runs `elabCommandTopLevel` itself,
+    -- so it is a "super-top-level" command. This is the only command that does this, so we just special case it here
+    -- rather than engineer a general solution.
+    unless (stx.find? (·.isOfKind ``Lean.guardMsgsCmd)).isSome do
```

Comment:
This means that linting for `#guard_msgs` or "`in`s that precede it" will become trickier, right?

E.g. the linters would only see `section` in
```lean
variable (x : Nat) in
open Lean in
#guard_msgs in
section
```
?

And something like
```lean
open Lean in
run_cmd
  let stx ← `(command| #guard_msgs in section)
  logInfo stx
```
would be completely unlinted?

Just to be explicit, I am happy with the change, just testing how far the consequences stretch!  :smile:

Reply:
You are correct, this change is at odds with the `#`-command linter. One idea that comes to mind is that we could have a core linter against `#guard_msgs` itself and enable it for mathlib, but I need to think about this more.

Reply:
This is not an issue for Mathlib: the `#`-command linter checks the kind of the command on the nos (no digging inside the syntax, as far as I remember).  Thus `#guard_msgs` actually appears as an `in` to that linter and `#guard_msgs` is *not* linted.

Reply:
I'm a bit confused about what you're saying — could you take a look at the log for the mathlib testing branch? There's an error in `test/HashCommandLinter.lean` where `#guard_msgs` is supposed to be linted. This PR makes it so that `#guard_msgs` never has linters run on it, and this causes that test to fail.

I'm not sure what you mean about `#guard_msgs` appearing as an `in`; the `in` is part of its syntax, it's not `Lean.Parser.Command.in`.

Reply:
Part of my confusion was due to the fact that the error in the tests is not visible in VSCode, but only with `lake test`.

The other confusion is that I remembered that `in` shielded the `#`-command linter, but really it is the doc-string that usually accompanies `#guard_msgs` that makes the linter ignore `#guard_msgs`.

Hence, as is, it is hard for `#guard_msgs` to get caught in the `#`-command linter:
```lean
import Mathlib.Tactic.Linter

--flagged
#guard_msgs in
run_cmd return

-- not flagged
/---/
#guard_msgs in
run_cmd return

-- not flagged
/-- info: 0 -/
#guard_msgs in
#eval 0
```
This behaviour was partly by design.  I would be happy to have `#guard_msgs` flagged more consistently as a `#`-command, but also to make it maintain "special status".

Reply:
Thanks for your feedback @adomani, it's helpful to know that you are ok with `#guard_msgs` being special like this. We can revisit being able to lint for `#guard_msgs` in mathlib if that ever comes up.

Would you mind fixing that first mathlib `HashCommandLinter.lean` test as a mathlib PR? That will help with the next nightly-testing.

Reply:
I'm going to fix the test in the same `lean-pr-testing-5644` branch, right?

Reply:
I don't think it's necessary to fix `lean-pr-testing-5644`. Maybe all we need to do is fix the nightly-testing branch of mathlib directly.

Pinging @jcommelin: My understanding is that in the next nightly, we'll have to fix the first test in HashCommandLinter.

Reply:
Ok, I have added the fix to `lean-pr-testing-5644` and I also removed the "hacky pairs" of `set_option ... false #guard_msgs in set_option ... true in ...`.

I can cherry-pick these changes to the appropriate nightly PR, though.

Reply:
Thanks @adomani!